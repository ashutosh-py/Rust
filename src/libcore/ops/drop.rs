/// Destructors for cleaning up resources
///
/// When a value is no longer needed, Rust will run a "destructor" on that value.
/// The most common way that a value is no longer needed is when it goes out of
/// scope. Destructors may still run in other circumstances, but we're going to
/// focus on scope for the examples here. To learn about some of those other cases,
/// please see [the reference] section on destructors.
///
/// [the reference]: https://doc.rust-lang.org/reference/destructors.html
///
/// The `Drop` trait provides a way to implement a custom destructor on a type.
/// Why would you want such a thing? Well, many types aren't only data: they
/// manage some sort of resource. That resource may be memory, it may be a file
/// descriptor, it may be a network socket. But when the type is no longer going
/// to be used, it should "clean up" that resource by freeing the memory or
/// closing the file or socket. This is the job of a destructor, and therefore
/// the job for `Drop::drop`.
///
/// ## Implementing `Drop`
///
/// When a value goes out of scope, it will call `Drop::drop` on that value. For example,
///
/// ```rust
/// struct HasDrop;
///
/// impl Drop for HasDrop {
///     fn drop(&mut self) {
///         println!("Dropping!");
///     }
/// }
///
/// fn main() {
///     let _x = HasDrop;
/// }
/// ```
///
/// Here, `main` will print `Dropping!`. In other words, the compiler generates code that
/// kind of looks like this:
///
/// ```rust,compile_fail,E0040
/// # struct HasDrop;
///
/// # impl Drop for HasDrop {
/// #     fn drop(&mut self) {
/// #         println!("Dropping!");
/// #     }
/// # }
/// fn main() {
///     let _x = HasDrop;
///
///     let mut x = _x;
///     Drop::drop(&mut x);
/// }
/// ```
///
/// As you can see, our custom implementation of `Drop` will be called at the end of `main`.
///
/// ## You cannot call `Drop::drop` yourself
///
/// Because the compiler automatically calls `Drop::drop`, you cannot call it yourself. This
/// would lead to "double drop", where `Drop::drop` is called twice on the same value. This
/// can lead to things like "double frees".
///
/// In other words, if you tried to write the code in the previous example with the explicit
/// call to `Drop::drop`, you'd get a compiler error.
///
/// If you'd like to call `drop` yourself, there is something you can do: call [`std::mem::drop`].
/// This function will drop its argument. For more, see its documentation.
///
/// [`std::mem::drop`]: ../../std/mem/fn.drop.html
///
/// ## `Drop` is recursive
///
/// If your type is something like a `struct` or `enum` that is an aggregate of other types, then
/// Rust will call `Drop::drop` on the type first, and then recursively on everything it contains.
/// For example:
///
/// ```rust
/// struct HasDrop;
///
/// impl Drop for HasDrop {
///     fn drop(&mut self) {
///         println!("Dropping HasDrop!");
///     }
/// }
///
/// struct HasTwoDrops {
///     one: HasDrop,
///     two: HasDrop,
/// }
///
/// impl Drop for HasTwoDrops {
///     fn drop(&mut self) {
///         println!("Dropping HasTwoDrops!");
///     }
/// }
///
/// fn main() {
///     let _x = HasTwoDrops { one: HasDrop, two: HasDrop };
/// }
/// ```
///
/// This will print
///
/// ```text
/// Dropping HasTwoDrops!
/// Dropping HasDrop!
/// Dropping HasDrop!
/// ```
///
/// Similarly, a slice will drop each element in order.
///
/// Which of our two `HasDrop` drops first, though? It's the same order that
/// they're declared: first `one`, then `two`. If you'd like to try this
/// yourself, you can modify `HasDrop` above to contain some data, like an
/// integer, and then use it in the `println!` inside of `Drop`. This behavior
/// is guaranteed by the language.
///
/// ## `Copy` and `Drop` are exclusive
///
/// You cannot implement both [`Copy`] and `Drop` on the same type. Types that
/// are `Copy` don't manage resources, and can be freely copied. As such, they
/// cannot have destructors.
///
/// [`Copy`]: ../../std/marker/trait.Copy.html
#[lang = "drop"]
#[stable(feature = "rust1", since = "1.0.0")]
pub trait Drop {
    /// Executes the destructor for this type.
    ///
    /// This method is called implicitly when the value goes out of scope,
    /// and cannot be called explicitly (this is compiler error [E0040]).
    /// However, the [`std::mem::drop`] function in the prelude can be
    /// used to call the argument's `Drop` implementation.
    ///
    /// When this method has been called, `self` has not yet been deallocated.
    /// That only happens after the method is over.
    /// If this wasn't the case, `self` would be a dangling reference.
    ///
    /// # Panics
    ///
    /// Given that a [`panic!`] will call `drop` as it unwinds, any [`panic!`]
    /// in a `drop` implementation will likely abort.
    ///
    /// [E0040]: ../../error-index.html#E0040
    /// [`panic!`]: ../macro.panic.html
    /// [`std::mem::drop`]: ../../std/mem/fn.drop.html
    #[stable(feature = "rust1", since = "1.0.0")]
    fn drop(&mut self);
}
