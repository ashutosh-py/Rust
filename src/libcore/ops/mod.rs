// Copyright 2012 The Rust Project Developers. See the COPYRIGHT
// file at the top-level directory of this distribution and at
// http://rust-lang.org/COPYRIGHT.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

//! Overloadable operators.
//!
//! Implementing these traits allows you to overload certain operators.
//!
//! Some of these traits are imported by the prelude, so they are available in
//! every Rust program. Only operators backed by traits can be overloaded. For
//! example, the addition operator (`+`) can be overloaded through the [`Add`]
//! trait, but since the assignment operator (`=`) has no backing trait, there
//! is no way of overloading its semantics. Additionally, this module does not
//! provide any mechanism to create new operators. If traitless overloading or
//! custom operators are required, you should look toward macros or compiler
//! plugins to extend Rust's syntax.
//!
//! Note that the `&&` and `||` operators short-circuit, i.e. they only
//! evaluate their second operand if it contributes to the result. Since this
//! behavior is not enforceable by traits, `&&` and `||` are not supported as
//! overloadable operators.
//!
//! Many of the operators take their operands by value. In non-generic
//! contexts involving built-in types, this is usually not a problem.
//! However, using these operators in generic code, requires some
//! attention if values have to be reused as opposed to letting the operators
//! consume them. One option is to occasionally use [`clone`].
//! Another option is to rely on the types involved providing additional
//! operator implementations for references. For example, for a user-defined
//! type `T` which is supposed to support addition, it is probably a good
//! idea to have both `T` and `&T` implement the traits [`Add<T>`][`Add`] and
//! [`Add<&T>`][`Add`] so that generic code can be written without unnecessary
//! cloning.
//!
//! # Examples
//!
//! This example creates a `Point` struct that implements [`Add`] and [`Sub`],
//! and then demonstrates adding and subtracting two `Point`s.
//!
//! ```rust
//! use std::ops::{Add, Sub};
//!
//! #[derive(Debug)]
//! struct Point {
//!     x: i32,
//!     y: i32,
//! }
//!
//! impl Add for Point {
//!     type Output = Point;
//!
//!     fn add(self, other: Point) -> Point {
//!         Point {x: self.x + other.x, y: self.y + other.y}
//!     }
//! }
//!
//! impl Sub for Point {
//!     type Output = Point;
//!
//!     fn sub(self, other: Point) -> Point {
//!         Point {x: self.x - other.x, y: self.y - other.y}
//!     }
//! }
//! fn main() {
//!     println!("{:?}", Point {x: 1, y: 0} + Point {x: 2, y: 3});
//!     println!("{:?}", Point {x: 1, y: 0} - Point {x: 2, y: 3});
//! }
//! ```
//!
//! See the documentation for each trait for an example implementation.
//!
//! The [`Fn`], [`FnMut`], and [`FnOnce`] traits are implemented by types that can be
//! invoked like functions. Note that [`Fn`] takes `&self`, [`FnMut`] takes `&mut
//! self` and [`FnOnce`] takes `self`. These correspond to the three kinds of
//! methods that can be invoked on an instance: call-by-reference,
//! call-by-mutable-reference, and call-by-value. The most common use of these
//! traits is to act as bounds to higher-level functions that take functions or
//! closures as arguments.
//!
//! Taking a [`Fn`] as a parameter:
//!
//! ```rust
//! fn call_with_one<F>(func: F) -> usize
//!     where F: Fn(usize) -> usize
//! {
//!     func(1)
//! }
//!
//! let double = |x| x * 2;
//! assert_eq!(call_with_one(double), 2);
//! ```
//!
//! Taking a [`FnMut`] as a parameter:
//!
//! ```rust
//! fn do_twice<F>(mut func: F)
//!     where F: FnMut()
//! {
//!     func();
//!     func();
//! }
//!
//! let mut x: usize = 1;
//! {
//!     let add_two_to_x = || x += 2;
//!     do_twice(add_two_to_x);
//! }
//!
//! assert_eq!(x, 5);
//! ```
//!
//! Taking a [`FnOnce`] as a parameter:
//!
//! ```rust
//! fn consume_with_relish<F>(func: F)
//!     where F: FnOnce() -> String
//! {
//!     // `func` consumes its captured variables, so it cannot be run more
//!     // than once
//!     println!("Consumed: {}", func());
//!
//!     println!("Delicious!");
//!
//!     // Attempting to invoke `func()` again will throw a `use of moved
//!     // value` error for `func`
//! }
//!
//! let x = String::from("x");
//! let consume_and_return_x = move || x;
//! consume_with_relish(consume_and_return_x);
//!
//! // `consume_and_return_x` can no longer be invoked at this point
//! ```
//!
//! [`Fn`]: trait.Fn.html
//! [`FnMut`]: trait.FnMut.html
//! [`FnOnce`]: trait.FnOnce.html
//! [`Add`]: trait.Add.html
//! [`Sub`]: trait.Sub.html
//! [`clone`]: ../clone/trait.Clone.html#tymethod.clone

#![stable(feature = "rust1", since = "1.0.0")]

mod markers;

pub use markers::Drop;

pub use num::{Add, Sub, Mul, Div, Rem, Neg};
pub use num::{AddAssign, SubAssign, MulAssign, DivAssign, RemAssign};

pub use bit::{Neg, BitAnd, BitOr, BitXor, Shl, Shr};
pub use bit::{BitAndAssign, BitOrAssign, BitXorAssign, ShlAssign, ShrAssign};

pub use range::{RangeFull, Range, RangeFrom, RangeTo};
pub use range::{RangeInclusive, RangeToInclusive};

#[stable(feature = "rust1", since = "1.0.0")]
