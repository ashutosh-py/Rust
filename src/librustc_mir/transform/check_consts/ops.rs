//! Concrete error types for all operations which may be invalid in a certain const context.

use rustc::mir;
use rustc::session::config::nightly_options;
use rustc::session::parse::feature_err;
use rustc::ty::{Ty, TyCtxt};
use rustc_errors::struct_span_err;
use rustc_hir::def_id::DefId;
use rustc_hir::Constness;
use rustc_span::symbol::sym;
use rustc_span::{Span, Symbol};

use super::{ConstKind, Item};

/// An operation that is not *always* allowed in a const context.
pub trait NonConstOp: std::fmt::Debug {
    /// Whether this operation can be evaluated by miri.
    const IS_SUPPORTED_IN_MIRI: bool = true;

    /// Returns the `Symbol` for the feature gate that would allow this operation, or `None` if
    /// such a feature gate does not exist.
    fn feature_gate() -> Option<Symbol> {
        None
    }

    /// Returns `true` if this operation is allowed in the given item.
    ///
    /// This check should assume that we are not in a non-const `fn`, where all operations are
    /// legal.
    fn is_allowed_in_item(&self, item: &Item<'_, '_>) -> bool {
        Self::feature_gate().map_or(false, |gate| feature_allowed(item, gate))
    }

    fn emit_error(&self, item: &Item<'_, '_>, span: Span) {
        let mut err = struct_span_err!(
            item.tcx.sess,
            span,
            E0019,
            "{} contains unimplemented expression type",
            item.const_kind()
        );
        if item.tcx.sess.teach(&err.get_code().unwrap()) {
            err.note(
                "A function call isn't allowed in the const's initialization expression \
                      because the expression's value must be known at compile-time.",
            );
            err.note(
                "Remember: you can't use a function call inside a const's initialization \
                      expression! However, you can use it anywhere else.",
            );
        }
        err.emit();
    }
}

#[derive(Debug)]
pub struct Abort;
impl NonConstOp for Abort {}

#[derive(Debug)]
pub enum ArithmeticOp {
    Binary(mir::BinOp),
    Unary(mir::UnOp),
}

impl From<mir::BinOp> for ArithmeticOp {
    fn from(op: mir::BinOp) -> Self {
        ArithmeticOp::Binary(op)
    }
}

impl From<mir::UnOp> for ArithmeticOp {
    fn from(op: mir::UnOp) -> Self {
        ArithmeticOp::Unary(op)
    }
}

#[derive(Debug)]
pub struct Arithmetic<'a>(pub ArithmeticOp, pub Ty<'a>);
impl NonConstOp for Arithmetic<'_> {
    fn is_allowed_in_item(&self, item: &Item<'_, '_>) -> bool {
        item.const_kind() != ConstKind::ConstFn || !min_const_fn_checks_enabled(item)
    }

    fn emit_error(&self, item: &Item<'_, '_>, span: Span) {
        let msg =
            format!("operations on `{:?}` are not allowed in a {}", self.1, item.const_kind());
        feature_err(&item.tcx.sess.parse_sess, sym::const_fn, span, &msg).emit()
    }
}

/// A `Downcast` projection.
#[derive(Debug)]
pub struct Downcast;
impl NonConstOp for Downcast {
    fn feature_gate() -> Option<Symbol> {
        Some(sym::const_if_match)
    }
}

/// A function call where the callee is a pointer.
#[derive(Debug)]
pub struct FnCallIndirect;
impl NonConstOp for FnCallIndirect {
    fn emit_error(&self, item: &Item<'_, '_>, span: Span) {
        item.tcx.sess.struct_span_err(span, "indirect function calls are not supported").emit()
    }
}

/// A function call where the callee is not marked as `const`.
#[derive(Debug)]
pub struct FnCallNonConst(pub DefId);
impl NonConstOp for FnCallNonConst {
    fn emit_error(&self, item: &Item<'_, '_>, span: Span) {
        let mut err = struct_span_err!(
            item.tcx.sess,
            span,
            E0015,
            "calls in {}s are limited to constant functions, \
             tuple structs and tuple variants",
            item.const_kind(),
        );
        err.emit();
    }
}

/// A function call where the callee is not a function definition or function pointer, e.g. a
/// closure.
///
/// This can be subdivided in the future to produce a better error message.
#[derive(Debug)]
pub struct FnCallOther;
impl NonConstOp for FnCallOther {
    const IS_SUPPORTED_IN_MIRI: bool = false;
}

/// A call to a `#[unstable]` const fn or `#[rustc_const_unstable]` function.
///
/// Contains the name of the feature that would allow the use of this function.
#[derive(Debug)]
pub struct FnCallUnstable(pub DefId, pub Symbol);
impl NonConstOp for FnCallUnstable {
    fn emit_error(&self, item: &Item<'_, '_>, span: Span) {
        let FnCallUnstable(def_id, feature) = *self;

        let mut err = item.tcx.sess.struct_span_err(
            span,
            &format!("`{}` is not yet stable as a const fn", item.tcx.def_path_str(def_id)),
        );
        if nightly_options::is_nightly_build() {
            err.help(&format!("add `#![feature({})]` to the crate attributes to enable", feature));
        }
        err.emit();
    }
}

#[derive(Debug)]
pub struct Generator;
impl NonConstOp for Generator {}

#[derive(Debug)]
pub struct HeapAllocation;
impl NonConstOp for HeapAllocation {
    const IS_SUPPORTED_IN_MIRI: bool = false;

    fn emit_error(&self, item: &Item<'_, '_>, span: Span) {
        let mut err = struct_span_err!(
            item.tcx.sess,
            span,
            E0010,
            "allocations are not allowed in {}s",
            item.const_kind()
        );
        err.span_label(span, format!("allocation not allowed in {}s", item.const_kind()));
        if item.tcx.sess.teach(&err.get_code().unwrap()) {
            err.note(
                "The value of statics and constants must be known at compile time, \
                 and they live for the entire lifetime of a program. Creating a boxed \
                 value allocates memory on the heap at runtime, and therefore cannot \
                 be done at compile time.",
            );
        }
        err.emit();
    }
}

#[derive(Debug)]
pub struct IfOrMatch;
impl NonConstOp for IfOrMatch {
    fn feature_gate() -> Option<Symbol> {
        Some(sym::const_if_match)
    }

    fn emit_error(&self, item: &Item<'_, '_>, span: Span) {
        // This should be caught by the HIR const-checker.
        item.tcx.sess.delay_span_bug(span, "complex control flow is forbidden in a const context");
    }
}

#[derive(Debug)]
pub struct InlineAsm;
impl NonConstOp for InlineAsm {}

#[derive(Debug)]
pub struct LiveDrop;
impl NonConstOp for LiveDrop {
    fn emit_error(&self, item: &Item<'_, '_>, span: Span) {
        struct_span_err!(
            item.tcx.sess,
            span,
            E0493,
            "destructors cannot be evaluated at compile-time"
        )
        .span_label(span, format!("{}s cannot evaluate destructors", item.const_kind()))
        .emit();
    }
}

#[derive(Debug)]
pub struct Loop;
impl NonConstOp for Loop {
    fn feature_gate() -> Option<Symbol> {
        Some(sym::const_loop)
    }

    fn emit_error(&self, item: &Item<'_, '_>, span: Span) {
        // This should be caught by the HIR const-checker.
        item.tcx.sess.delay_span_bug(span, "complex control flow is forbidden in a const context");
    }
}

#[derive(Debug)]
pub struct CellBorrow;
impl NonConstOp for CellBorrow {
    fn emit_error(&self, item: &Item<'_, '_>, span: Span) {
        struct_span_err!(
            item.tcx.sess,
            span,
            E0492,
            "cannot borrow a constant which may contain \
            interior mutability, create a static instead"
        )
        .emit();
    }
}

#[derive(Debug)]
pub struct MutBorrow;
impl NonConstOp for MutBorrow {
    fn feature_gate() -> Option<Symbol> {
        Some(sym::const_mut_refs)
    }

    fn emit_error(&self, item: &Item<'_, '_>, span: Span) {
        let mut err = feature_err(
            &item.tcx.sess.parse_sess,
            sym::const_mut_refs,
            span,
            &format!("mutable references in a {} are unstable", item.const_kind()),
        );
        err.span_label(span, format!("{}s require immutable values", item.const_kind()));
        if item.tcx.sess.teach(&err.get_code().unwrap()) {
            err.note(
                "References in statics and constants may only refer \
                      to immutable values.\n\n\
                      Statics are shared everywhere, and if they refer to \
                      mutable data one might violate memory safety since \
                      holding multiple mutable references to shared data \
                      is not allowed.\n\n\
                      If you really want global mutable state, try using \
                      static mut or a global UnsafeCell.",
            );
        }
        err.emit();
    }
}

#[derive(Debug)]
pub struct MutDeref;
impl NonConstOp for MutDeref {
    fn feature_gate() -> Option<Symbol> {
        Some(sym::const_mut_refs)
    }
}

#[derive(Debug)]
pub struct Panic;
impl NonConstOp for Panic {
    fn feature_gate() -> Option<Symbol> {
        Some(sym::const_panic)
    }

    fn emit_error(&self, item: &Item<'_, '_>, span: Span) {
        feature_err(
            &item.tcx.sess.parse_sess,
            sym::const_panic,
            span,
            &format!("panicking in {}s is unstable", item.const_kind()),
        )
        .emit();
    }
}

#[derive(Debug)]
pub struct RawPtrComparison;
impl NonConstOp for RawPtrComparison {
    fn feature_gate() -> Option<Symbol> {
        Some(sym::const_compare_raw_pointers)
    }

    fn emit_error(&self, item: &Item<'_, '_>, span: Span) {
        feature_err(
            &item.tcx.sess.parse_sess,
            sym::const_compare_raw_pointers,
            span,
            &format!("comparing raw pointers inside {}", item.const_kind()),
        )
        .emit();
    }
}

#[derive(Debug)]
pub struct RawPtrDeref;
impl NonConstOp for RawPtrDeref {
    fn feature_gate() -> Option<Symbol> {
        Some(sym::const_raw_ptr_deref)
    }

    fn emit_error(&self, item: &Item<'_, '_>, span: Span) {
        feature_err(
            &item.tcx.sess.parse_sess,
            sym::const_raw_ptr_deref,
            span,
            &format!("dereferencing raw pointers in {}s is unstable", item.const_kind()),
        )
        .emit();
    }
}

#[derive(Debug)]
pub struct RawPtrToIntCast;
impl NonConstOp for RawPtrToIntCast {
    fn feature_gate() -> Option<Symbol> {
        Some(sym::const_raw_ptr_to_usize_cast)
    }

    fn emit_error(&self, item: &Item<'_, '_>, span: Span) {
        feature_err(
            &item.tcx.sess.parse_sess,
            sym::const_raw_ptr_to_usize_cast,
            span,
            &format!("casting pointers to integers in {}s is unstable", item.const_kind()),
        )
        .emit();
    }
}

#[derive(Debug)]
pub struct FnPtr;
impl NonConstOp for FnPtr {
    fn is_allowed_in_item(&self, item: &Item<'_, '_>) -> bool {
        item.const_kind() != ConstKind::ConstFn
            || !min_const_fn_checks_enabled(item)
            || item.tcx.has_attr(item.def_id, sym::rustc_allow_const_fn_ptr)
    }

    fn emit_error(&self, item: &Item<'_, '_>, span: Span) {
        feature_err(
            &item.tcx.sess.parse_sess,
            sym::const_fn,
            span,
            "function pointers in const fn are unstable",
        )
        .emit()
    }
}

/// See [#64992].
///
/// [#64992]: https://github.com/rust-lang/rust/issues/64992
#[derive(Debug)]
pub struct UnsizingCast;
impl NonConstOp for UnsizingCast {
    fn is_allowed_in_item(&self, item: &Item<'_, '_>) -> bool {
        item.const_kind() != ConstKind::ConstFn || !min_const_fn_checks_enabled(item)
    }

    fn emit_error(&self, item: &Item<'_, '_>, span: Span) {
        feature_err(
            &item.tcx.sess.parse_sess,
            sym::const_fn,
            span,
            "unsizing casts are not allowed in const fn",
        )
        .emit()
    }
}

#[derive(Debug)]
pub struct ImplTrait;
impl NonConstOp for ImplTrait {
    fn is_allowed_in_item(&self, item: &Item<'_, '_>) -> bool {
        item.const_kind() != ConstKind::ConstFn || !min_const_fn_checks_enabled(item)
    }

    fn emit_error(&self, item: &Item<'_, '_>, span: Span) {
        feature_err(
            &item.tcx.sess.parse_sess,
            sym::const_fn,
            span,
            "`impl Trait` in const fn is unstable",
        )
        .emit()
    }
}

#[derive(Debug)]
pub struct TraitBound(pub Constness);
impl NonConstOp for TraitBound {
    fn is_allowed_in_item(&self, item: &Item<'_, '_>) -> bool {
        if item.const_kind() != ConstKind::ConstFn || !min_const_fn_checks_enabled(item) {
            return true;
        }

        // Allow `T: ?const Trait`
        if self.0 == Constness::NotConst {
            feature_allowed(item, sym::const_trait_bound_opt_out)
        } else {
            false
        }
    }

    fn emit_error(&self, item: &Item<'_, '_>, span: Span) {
        feature_err(
            &item.tcx.sess.parse_sess,
            sym::const_fn,
            span,
            "trait bounds other than `Sized` on const fn parameters are unstable",
        )
        .emit()
    }
}

/// An access to a (non-thread-local) `static`.
#[derive(Debug)]
pub struct StaticAccess;
impl NonConstOp for StaticAccess {
    fn is_allowed_in_item(&self, item: &Item<'_, '_>) -> bool {
        item.const_kind().is_static()
    }

    fn emit_error(&self, item: &Item<'_, '_>, span: Span) {
        let mut err = struct_span_err!(
            item.tcx.sess,
            span,
            E0013,
            "{}s cannot refer to statics",
            item.const_kind()
        );
        err.help(
            "consider extracting the value of the `static` to a `const`, and referring to that",
        );
        if item.tcx.sess.teach(&err.get_code().unwrap()) {
            err.note(
                "`static` and `const` variables can refer to other `const` variables. \
                    A `const` variable, however, cannot refer to a `static` variable.",
            );
            err.help("To fix this, the value can be extracted to a `const` and then used.");
        }
        err.emit();
    }
}

/// An access to a thread-local `static`.
#[derive(Debug)]
pub struct ThreadLocalAccess;
impl NonConstOp for ThreadLocalAccess {
    const IS_SUPPORTED_IN_MIRI: bool = false;

    fn emit_error(&self, item: &Item<'_, '_>, span: Span) {
        struct_span_err!(
            item.tcx.sess,
            span,
            E0625,
            "thread-local statics cannot be \
            accessed at compile-time"
        )
        .emit();
    }
}

#[derive(Debug)]
pub struct UnionAccess;
impl NonConstOp for UnionAccess {
    fn is_allowed_in_item(&self, item: &Item<'_, '_>) -> bool {
        // Union accesses are stable in all contexts except `const fn`.
        item.const_kind() != ConstKind::ConstFn || feature_allowed(item, sym::const_fn_union)
    }

    fn feature_gate() -> Option<Symbol> {
        Some(sym::const_fn_union)
    }

    fn emit_error(&self, item: &Item<'_, '_>, span: Span) {
        feature_err(
            &item.tcx.sess.parse_sess,
            sym::const_fn_union,
            span,
            "unions in const fn are unstable",
        )
        .emit();
    }
}

/// Returns `true` if the constness of this item is not stabilized, either because it is declared
/// with `#[rustc_const_unstable]` or because the item itself is unstable.
fn is_declared_const_unstable(tcx: TyCtxt<'tcx>, def_id: DefId) -> bool {
    tcx.lookup_const_stability(def_id).map_or(false, |stab| stab.level.is_unstable())
        || tcx.lookup_stability(def_id).map_or(false, |stab| stab.level.is_unstable())
}

/// Returns `true` if the feature with the given gate is allowed within this const context.
fn feature_allowed(item: &Item<'_, '_>, feature_gate: Symbol) -> bool {
    let Item { tcx, def_id, .. } = *item;

    // All features require that the corresponding gate be enabled,
    // even if the function has `#[allow_internal_unstable(the_gate)]`.
    if !tcx.features().enabled(feature_gate) {
        return false;
    }

    // If this crate is not using stability attributes, or this function is not claiming to be a
    // stable `const fn`, that is all that is required.
    if !tcx.features().staged_api {
        return true;
    }

    if is_declared_const_unstable(item.tcx, item.def_id) {
        return true;
    }

    // However, we cannot allow stable `const fn`s to use unstable features without an explicit
    // opt-in via `allow_internal_unstable`.
    rustc_attr::allow_internal_unstable(&tcx.get_attrs(def_id), &tcx.sess.diagnostic())
        .map_or(false, |mut features| features.any(|name| name == feature_gate))
}

fn min_const_fn_checks_enabled(item: &Item<'_, '_>) -> bool {
    assert_eq!(item.const_kind(), ConstKind::ConstFn);

    if item.tcx.features().staged_api {
        // All functions except for unstable ones need to pass the min const fn checks. This
        // includes private functions that are not marked unstable.
        !is_declared_const_unstable(item.tcx, item.def_id)
    } else {
        // Crates that are not using stability attributes can use `#![feature(const_fn)]` to opt out of
        // the min_const_fn checks.
        !item.tcx.features().const_fn
    }
}
