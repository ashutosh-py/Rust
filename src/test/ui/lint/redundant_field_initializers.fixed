// run-rustfix
#![deny(redundant_field_initializers)]
#![allow(dead_code, unused_variables)]

use std::ops::{Range, RangeFrom, RangeInclusive, RangeTo, RangeToInclusive};

mod foo {
    pub const BAR: u8 = 0;
}

struct Person {
    gender: u8,
    age: u8,
    name: u8,
    buzz: u64,
    foo: u8,
}

fn main() {
    let gender: u8 = 42;
    let age = 0;
    let fizz: u64 = 0;
    let name: u8 = 0;

    let me = Person {
        gender,
        //~^ ERROR redundant field names
        age,
        //~^ ERROR redundant field names

        name,          //should be ok
        buzz: fizz,    //should be ok
        foo: foo::BAR, //should be ok
    };

    // Range expressions
    let (start, end) = (0, 0);

    let _ = start..;
    let _ = ..end;
    let _ = start..end;

    let _ = ..=end;
    let _ = start..=end;

    // Issue #2799
    let _: Vec<_> = (start..end).collect();

    // hand-written Range family structs are linted
    let _ = RangeFrom { start }; //~ ERROR redundant field names
    let _ = RangeTo { end }; //~ ERROR redundant field names
    let _ = Range {
        start, //~ ERROR redundant field names
        end //~ ERROR redundant field names
    };
    let _ = RangeInclusive::new(start, end);
    let _ = RangeToInclusive { end }; //~ ERROR redundant field names
}

fn issue_3476() {
    fn foo<T>() {}

    struct S {
        foo: fn(),
    }

    S { foo: foo::<i32> };
}
