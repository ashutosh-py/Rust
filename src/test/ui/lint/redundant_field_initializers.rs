// run-rustfix
#![deny(redundant_field_initializers)]
#![allow(dead_code, unused_variables)]

use std::ops::{Range, RangeFrom, RangeInclusive, RangeTo, RangeToInclusive};

mod foo {
    pub const BAR: u8 = 0;
}

struct Person {
    gender: u8,
    age: u8,
    name: u8,
    buzz: u64,
    foo: u8,
}

fn main() {
    let gender: u8 = 42;
    let age = 0;
    let fizz: u64 = 0;
    let name: u8 = 0;

    let me = Person {
        gender: gender,
        //~^ ERROR redundant field names
        age: age,
        //~^ ERROR redundant field names

        name,          //should be ok
        buzz: fizz,    //should be ok
        foo: foo::BAR, //should be ok
    };

    // Range expressions
    let (start, end) = (0, 0);

    let _ = start..;
    let _ = ..end;
    let _ = start..end;

    let _ = ..=end;
    let _ = start..=end;

    // Issue #2799
    let _: Vec<_> = (start..end).collect();

    // hand-written Range family structs are linted
    let _ = RangeFrom { start: start }; //~ ERROR redundant field names
    let _ = RangeTo { end: end }; //~ ERROR redundant field names
    let _ = Range {
        start: start, //~ ERROR redundant field names
        end: end //~ ERROR redundant field names
    };
    let _ = RangeInclusive::new(start, end);
    let _ = RangeToInclusive { end: end }; //~ ERROR redundant field names
}

fn issue_3476() {
    fn foo<T>() {}

    struct S {
        foo: fn(),
    }

    S { foo: foo::<i32> };
}
