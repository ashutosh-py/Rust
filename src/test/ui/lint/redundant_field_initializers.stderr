error: redundant field names in struct initialization
  --> $DIR/redundant_field_initializers.rs:26:9
   |
LL |         gender: gender,
   |         ^^^^^^^^^^^^^^ help: replace it with: `gender`
   |
note: the lint level is defined here
  --> $DIR/redundant_field_initializers.rs:2:9
   |
LL | #![deny(redundant_field_initializers)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: redundant field names in struct initialization
  --> $DIR/redundant_field_initializers.rs:28:9
   |
LL |         age: age,
   |         ^^^^^^^^ help: replace it with: `age`

error: redundant field names in struct initialization
  --> $DIR/redundant_field_initializers.rs:50:25
   |
LL |     let _ = RangeFrom { start: start };
   |                         ^^^^^^^^^^^^ help: replace it with: `start`

error: redundant field names in struct initialization
  --> $DIR/redundant_field_initializers.rs:51:23
   |
LL |     let _ = RangeTo { end: end };
   |                       ^^^^^^^^ help: replace it with: `end`

error: redundant field names in struct initialization
  --> $DIR/redundant_field_initializers.rs:53:9
   |
LL |         start: start,
   |         ^^^^^^^^^^^^ help: replace it with: `start`

error: redundant field names in struct initialization
  --> $DIR/redundant_field_initializers.rs:54:9
   |
LL |         end: end
   |         ^^^^^^^^ help: replace it with: `end`

error: redundant field names in struct initialization
  --> $DIR/redundant_field_initializers.rs:57:32
   |
LL |     let _ = RangeToInclusive { end: end };
   |                                ^^^^^^^^ help: replace it with: `end`

error: aborting due to 7 previous errors

